### PR TITLE
Return a 404 from `/state` for an outlier

### LINKED ISSUE #12074: `/_matrix/federation/v1/state` returns a 500 error if called for an outlier
```
2022-02-24 09:59:19,270 - synapse.http.server - 100 - ERROR - GET-281 - Failed handle request via 'FederationStateV1Servlet': <SynapseRequest at 0x7f683468c0a0 method='GET' uri='/_matrix/federation/v1/state/%21ItGBZyOeGywOWwJbSK%3Alocalhost%3A8840?event_id=%242DMzm1OALGG81SL30L-ArnxcWKfwrlvkv-bqehxWNdk' clientproto='HTTP/1.1' site='8849'>
Traceback (most recent call last):
  File "/venv/lib/python3.9/site-packages/twisted/internet/defer.py", line 1660, in _inlineCallbacks
    result = current_context.run(gen.send, result)
StopIteration: 0

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/venv/lib/python3.9/site-packages/twisted/internet/defer.py", line 1660, in _inlineCallbacks
    result = current_context.run(gen.send, result)
StopIteration: {'ed25519:a_VhKa': FetchKeyResult(verify_key=<nacl.signing.VerifyKey object at 0x7f6834b83550>, valid_until_ts=1645782543370)}

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/venv/lib/python3.9/site-packages/synapse/http/server.py", line 269, in _async_render_wrapper
    callback_return = await self._async_render(request)
  File "/venv/lib/python3.9/site-packages/synapse/http/server.py", line 471, in _async_render
    callback_return = await raw_callback_return
  File "/venv/lib/python3.9/site-packages/synapse/federation/transport/server/_base.py", line 323, in new_func
    response = await func(
  File "/venv/lib/python3.9/site-packages/synapse/federation/transport/server/federation.py", line 160, in on_GET
    return await self.handler.on_room_state_request(
  File "/venv/lib/python3.9/site-packages/synapse/federation/federation_server.py", line 535, in on_room_state_request
    await self._state_resp_cache.wrap(
  File "/venv/lib/python3.9/site-packages/synapse/util/caches/response_cache.py", line 250, in wrap
    return await make_deferred_yieldable(entry.result.observe())
  File "/venv/lib/python3.9/site-packages/twisted/internet/defer.py", line 1660, in _inlineCallbacks
    result = current_context.run(gen.send, result)
  File "/venv/lib/python3.9/site-packages/synapse/util/caches/response_cache.py", line 246, in cb
    return await callback(*args, **kwargs)
  File "/venv/lib/python3.9/site-packages/synapse/federation/federation_server.py", line 581, in _on_context_state_request_compute
    pdus: Iterable[EventBase] = await self.handler.get_state_for_pdu(
  File "/venv/lib/python3.9/site-packages/synapse/handlers/federation.py", line 1015, in get_state_for_pdu
    del results[(event.type, event.state_key)]
KeyError: ('m.room.member', '@anon-20220224_094858-726:localhost:8881')

### PR DESCRIPTION
**Pull request by richvdh**
Fixes #12074

The main fix here is the `if event.internal_metadata.outlier` condition, which makes sure we return a sensible result in this situation, but I suggest reviewing commit-by-commit as there are a few other cleanups while we're here.

### CODE REVIEW

**Comment by erikjohnston** [comment]
I assume since this is a draft it's not ready for review?

**Comment by richvdh** [comment]
> I assume since this is a draft it's not ready for review?

correct. I'm working on getting the tests passing (and getting distracted by things like #12089 and #12109).  It shouldn't be showing up in the review queue yet?

**Comment by richvdh** [comment]
(finally! the tests pass!)

File: synapse/handlers/federation.py [hunk 1]
```diff
@@ -963,14 +963,17 @@ async def get_state_ids_for_pdu(self, room_id: str, event_id: str) -> List[str]:
 
         state_map = next(iter(state_groups.values()))
 
-        if event.is_state():
-            # Get previous state
+        # we need the state *before* event.
+        state_key = event.get_state_key()
+        if (
+            state_key is not None
+            and state_map.get((event.type, state_key)) == event.event_id
```
Comment (clokep) [inline synapse/handlers/federation.py:970]: I'm struggling to see why this extra check against the event ID; could it be that sometimes `state_map` might already have the previous state?

Comment (richvdh) [reply synapse/handlers/federation.py:970]: hrm, good question.

`state_map` certainly won't include `event` if `event` was rejected... but it's not clear to me what the behaviour of this method is when `event` is rejected anyway.

I guess I'll have to poke at this a bit more to understand it.

Comment (richvdh) [reply synapse/handlers/federation.py]: turns out that we never get this far if the event was rejected (i'll put a sytest up for that in a bit), so I really can't think of any reason `state_map` wouldn't include `event`. Accordingly, I've replaced this with an `assert`.

**Review by clokep — APPROVED** [review]
LGTM If CI is happy.

File: synapse/handlers/federation.py (hunk 1 above)
Comment (clokep) [reply synapse/handlers/federation.py]: Thanks for investigating! The comments are clear now!

**Pull request merged**